From efced8ad47f43a18bf8c992a61ee29f6ed031aba Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 18 Aug 2014 23:56:40 +0800
Subject: [PATCH] bfs: Make grq cacheline aligned.

---
 kernel/sched/bfs.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index 2543a83..294a002 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -224,7 +224,11 @@ static struct root_domain def_root_domain;
 #endif /* CONFIG_SMP */
 
 /* There can be only one */
-static struct global_rq grq;
+#ifdef CONFIG_SMP
+static struct global_rq grq ____cacheline_aligned_in_smp;
+#else
+static struct global_rq grq ____cacheline_aligned;
+#endif
 
 static DEFINE_MUTEX(sched_hotcpu_mutex);
 
-- 
2.1.1

