From 5de62b5f4f3ca8caa1803a306a35c1740b8171fd Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 15 Oct 2014 10:32:11 +0800
Subject: [PATCH] bfs: Refactor needs_other_cpu().

---
 kernel/sched/bfs.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index e80e9a11..44140ea 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -1343,9 +1343,7 @@ can_preempt(struct task_struct *p, int prio, u64 deadline)
  */
 static inline bool needs_other_cpu(struct task_struct *p, int cpu)
 {
-	if (unlikely(!cpu_isset(cpu, p->cpus_allowed)))
-		return true;
-	return false;
+	return !cpumask_test_cpu(cpu, &p->cpus_allowed);
 }
 
 /*
-- 
2.1.1

